clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -1568,6 +1571,17 @@ public void handleResponse(AbstractResponse response) {
             Errors error = endTxnResponse.error();
 
             if (error == Errors.NONE) {
+                // For transaction version 5+, the broker includes the producerId and producerEpoch in the EndTxnResponse.
+                // KIP-890 Part 2 mandates bumping the epoch after every transaction. If the epoch overflows,
+                // a new producerId is returned with epoch set to 0.
+                if (isTransactionV2Enabled) {
```

Comment:
I think this condition should be based on the response version

Reply:
The isTransactionV2Enabled is locked within a transaction's life cycle. I wonder if we should abort the transaction if  isTransactionV2Enabled=true but endTxnResponse version<5.

Reply:
There is no reason to do this if we already received the response.

Reply:
My understanding is that downgrade path would work as following:
1. Client may see the that v2 moved back to v1, but continues current transaction with v2.
2. Brokers will keep receiving new RPC versions until the transaction ends.
3. Client switches to v1 at the beginning of the next transaction and continues further with v1.

During step 2 one of two things could happen:
a. The broker image still supports new protocol (even though it see v1 as a TV in the metadata). The broker will honor the new protocol (this behavior is anyway required for the upgrade path, because a broker may see v2 metadata update later than the client).
b. The broker is downgraded to an old image.  The client should get a version mismatch error, which would be treated as fatal and require producer close.

We may need to add a test to test for a case 2b.

Reply:
Thanks for enumerating it out Artem. I agree with this.

Reply:
I understand that we would rather use the response version to be consistent with the server but I'm trying to understand when we would be in a situation where TV2 = true but the EndTxnResponse < 5 or vice versa where TV2 = false and EndTxnResponse > 5. Afaik it's always a 1:1 mapping but I might be missing something. Unless it's possible to change the TV mid transaction. Is there some edge case we want to avoid by using the response version instead of the flag?

Reply:
TV can be updated dynamically. So yes, it could be updated mid-transaction. And no, we don't want to check the TV since the response tells us the state the server already committed, and we don't want the client to be wrong and not be able to proceed since it will be fenced.

Reply:
> TV2 = true but the EndTxnResponse < 5
This can happen if
- TV2 = true when the transaction starts. (The client will check TV2 = true until the next transaction)
- The broker has an image downgrade which does not support EndTxn version 5 anymore
- The client updates the Api versions.
- The client sends the EndTxn version 4 and gets the responses of version 4.
It is super rare and like Justine mentioned, we can just use the EndTxn response to determine the action.

Reply:
@artemlivshits was telling me that we don't want to allow TV to be updated mid transaction, is that still the case?

Reply:
We don't want to change what the producer is doing mid transaction. We can not control what a user does -- they can change the TV at any time. We want them to be able to upgrade TV without having to stop all transactions.

Reply:
For this part
> The client sends the EndTxn version 4

Do we not force sending version 5? I think we may want to do that and get UnsupportedVersionException rather than switch the protocol mid transaction.

Reply:
> Do we not force sending version 5? I think we may want to do that and get UnsupportedVersionException

The desired behavior is for this^^ to happen.  If TV is downgraded in the middle of transaction, we don't change `isTransactionV2Enabled` until the end and continue using new protocol and new RPC versions.  If the broker image hasn't downgraded yet, then it'll just work.  If the broker has managed to downgrade, then we'd get UnsuppotedVersionException, so we should never get an older version response when `isTransactionV2Enabled == true`.  There are other ways to implement downgrade, but this one is simple to prove correctness (for example, we don't need to consider all the combinations of cases when protocol gets switched in the middle and we could send mixed versions to different brokers and prove that all combinations work correctly).

Reply:
> so we should never get an older version response when isTransactionV2Enabled == true
Makes sense, this was my understanding as well, thanks for clarifying. So essentially it shouldn't be possible but just to be safe we're gonna use the response version to double check if we can bump the epoch instead of relying on the flag since that behavior could be subject to change and cause problems right?